lib/Parse/ParsePattern.cpp

```diff
@@ -672,26 +703,16 @@ Parser::parseFunctionSignature(Identifier SimpleName,
   }
 
   // Check for 'throws' and 'rethrows' after the type and correct it.
-  if (!throwsLoc.isValid()) {
-    if (Tok.is(tok::kw_throws)) {
-      throwsLoc = consumeToken();
-    } else if (Tok.is(tok::kw_rethrows)) {
-      throwsLoc = consumeToken();
-      rethrows = true;
-    }
-
-    if (throwsLoc.isValid()) {
-      assert(arrowLoc.isValid());
-      assert(retType);
-      auto diag = rethrows ? diag::rethrows_after_function_result
-                           : diag::throws_after_function_result;
-      SourceLoc typeEndLoc = Lexer::getLocForEndOfToken(SourceMgr,
-                                                        retType->getEndLoc());
-      SourceLoc throwsEndLoc = Lexer::getLocForEndOfToken(SourceMgr, throwsLoc);
-      diagnose(Tok, diag)
-        .fixItInsert(arrowLoc, rethrows ? "rethrows " : "throws ")
-        .fixItRemoveChars(typeEndLoc, throwsEndLoc);
-    }
+  auto diagOpt = diagnoseInvalidThrows();
+  if (diagOpt) {
```

Comment:
`if (auto diagOpt = diagnoseInvalidThrows())`?  Won't unwrap the optional, but at least would reduce the scope of the variable.